File: dynamic_programming/01_matrix01/matrix01.py
import numpy as np
from collections import deque
from typing import Deque, Tuple


class Solution:
    def update_matrix(self, mat: np.ndarray) -> np.ndarray:
        if not self._check_constraints(mat):
            raise ValueError('Constraints violated')

        row, col = mat.shape
        dist = np.ones_like(mat) * np.iinfo(np.int32).max
        q: Deque[Tuple[int, int]] = deque()
        # Init as a root for BFS every cell that is 0
        for i in range(row):
            for j in range(col):
                if (mat[i, j] == 0):
                    dist[i, j] = 0
                    q.append((i, j))
        dirs = [(1, 0), (-1, 0), (0, 1), (0, -1)]
        while q:
            id = q.popleft()
            for dir in dirs:
                x, y = id[0] + dir[0], id[1] + dir[1]
                if x >= 0 and x < row and y >= 0 and y < col:
                    if dist[x, y] > dist[id] + 1:
                        dist[x, y] = dist[id] + 1
                        q.append((x, y))

        return dist

    def _check_constraints(self, mat: np.ndarray) -> bool:
        m = mat.shape[0]
        if m == 0 or m > 10**4:
            return False
        zero_cnt = 0
        for row in mat:
            if row.shape[0] == 0 or row.shape[0] > 10**4:
                return False
            for num in row:
                if not (num == 0 or num == 1):
                    return False
                if num == 0:
                    zero_cnt += 1

        if not zero_cnt:
            return False

        return True

File: dynamic_programming/01_matrix01/matrix01.cpp
#include <iostream>
#include <stdexcept>
#include <cmath>
#include <cstdint>
#include <algorithm>
#include <vector>
#include <queue>

using namespace std;


class Solution {
public:
    vector<vector<int>> updateMatrix(const vector<vector<int>>& mat) {
        if (!checkConstraints(mat)) {
            throw std::runtime_error("Constraints violated");
        }
        int row = mat.size(), col = mat[0].size();
        vector<vector<int>> dist(row, vector<int> (col, INT32_MAX));
        queue<pair<int, int>> q;
        // Init as a root for BFS every cell that is 0
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                if (mat[i][j] == 0) {
                    dist[i][j] = 0;
                    q.push({i, j});
                }
            }
        }
        vector<pair<int, int>> dirs = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        // BFS
        while (!q.empty()) {
            auto id = q.front();
            q.pop();
            for (auto& dir : dirs) {
                int x = id.first + dir.first, y = id.second + dir.second;
                if (x >= 0 && x < row && y >= 0 && y < col) {
                    if (dist[x][y] > dist[id.first][id.second] + 1) {
                        dist[x][y] = dist[id.first][id.second] + 1;
                        q.push({x, y});
                    }
                }
            }
        }
        return dist;
    }

private:
    // Helper function to check the constraints
    bool checkConstraints(const vector<vector<int>>& mat) {
        int m = mat.size();
        if (m == 0 || m > pow(10, 4)) return false;
        int zeroCnt = 0;
        for (const auto& row : mat) {
            if (row.size() == 0 || row.size() > pow(10, 4)) return false;
            if (row.size() * m == 0 || row.size() * m > pow(10, 4)) return false;
            for (int num : row) {
                if (num == 0) {
                    zeroCnt++;
                }
                if (num != 0 && num != 1) return false;
            }
        }
        if (!zeroCnt) return false;

        return true;
    }
};
